### DISCUSSION TITLE
Default to bash for all recipes ?

### DISCUSSION THREAD

**Discussion by mvuidev** [casey/just#1353] [Q&A] [answered]
Hello,

All my recipes start with :

```bash
#!/bin/bash
set -euo pipefail
```

It's a bit redundant, so is there a way to make it global for all the recipes ?

Thanks !

**Comment by casey** [comment, marked as answer]
Can you give an example justfile showing why you need this?

**Reply by mvuidev** [reply]
Sure, but I don't think what I want to do is possible :)

Basically, I only use bash in my recipes, so for example : 

```bash
func_1:
    #!/bin/bash
    set -euo pipefail

    echo "I am func_1"

func_2:
    #!/bin/bash
    set -euo pipefail

    echo "I am func_2"
```

And so on.
I was just wondering if there was a way to not type : 

```bash
#!/bin/bash
set -euo pipefail
```

In every recipe ?

**Reply by casey** [reply]
Can you provide more details about *why* you use bash in your recipes? I.e. what feature are you using?

**Reply by mvuidev** [reply]
Well, I have multiple bash (only) scripts that I converted to Justfiles (system and project dev scripts).
In some recipes I have to define variables, conditions, run ssh multiline commands, etc. but coundn't make this work without the shebang.

Examples :

```bash
recipe:
    #!/bin/bash
    set -euo pipefail

    # Variable
    var=$(command)
    echo $var

    # Condition
    if [[ -f "file.ext" ]]; then
        some_function
    fi

    # SSH commands
    ssh host "bash -s" <<-'EOF'
        some
        commands
    EOF
```

I could use the shebang only in recipes that requires it but you know, I thought "why not use it everywhere..." :roll_eyes:

**Reply by casey** [reply]
Gotcha! I just wanted to make sure I understood why you were using bash, in case maybe there was another way to solve your problem. But you're right that things like variables and multi-line constructs require a `#!` recipe. Currently there's no way to do this. You can customize the shell that runs linewise recipes (recipes without a `#!`) but there's no way to make bash the default for shebang recipes, or to make all recipes shebang recipes.

**Reply by mvuidev** [reply]
Thank you for the answer, that's what I thought. Not a big deal, I'll continue to use the shebang :) And Just is really nice, good job ! I'm missing some features, but I hope it'll improve. I'll probably open another discussion.

**Comment by runeimp** [comment]
That could get dicey real quick as you could easily end up having a script that you need to add to your `Justfile` that needs to used other Bash settings. At that point how do you manage the situation? Have `just` scan the recipe and disable the default setting if `set` is used in any way within the script? That would be bad if your script switches back and forth with different `set` parameters, and expects _NOT_ to start with `set -euo pipefail`. Also, what if others need to use your `Justfile`? Someone who is a competent Bash scripter could easily miss the special setting a the top of the Justfile which they may not even understand if they saw it depending on how it's worded.

**Reply by mvuidev** [reply]
Yeah you're right, maybe it's not a great idea after all. Not concerned about the last point cause I'm the only one using the justfiles, but anyway, should be safer to stick with shebangs where it's needed. Thanks for your comment !
